pkg/ipam/poolallocator/allocator.go

```diff
@@ -0,0 +1,254 @@
+// Copyright 2021 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package poolallocator
+
+import (
+	"context"
+	"fmt"
+	"net"
+
+	"antrea.io/antrea/pkg/apis/crd/v1alpha2"
+	crdclientset "antrea.io/antrea/pkg/client/clientset/versioned"
+	"antrea.io/antrea/pkg/ipam/ipallocator"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+)
+
+// IPPoolAllocator is responsible for allocating IPs from IP set defined in IPPool CRD.
+// The will update CRD usage accordingly.
+type IPPoolAllocator struct {
+	// Name of IP Pool custom resource
+	ipPoolName string
+
+	// crd client to access the pool
+	crdClient crdclientset.Interface
+}
+
+// NewIPPoolAllocator creates an IPPoolAllocator based on the provided IP pool.
+func NewIPPoolAllocator(poolName string, client crdclientset.Interface) (*IPPoolAllocator, error) {
+
+	allocator := &IPPoolAllocator{
+		ipPoolName: poolName,
+		crdClient:  client,
+	}
+	return allocator, nil
+}
+
+// initAllocatorList reads IP Pool status and initializes a list of allocators based on
+// IP Pool spec and state of allocation recorded in the status
+func (a *IPPoolAllocator) initIPAllocators(ipPool *v1alpha2.IPPool) (ipallocator.MultiIPAllocator, error) {
```

Comment:
Is it possible not to init IPAllocators on every allocate/release request? Or you think this method is good enough when an IPPool CRD won't contain too many IPRanges.

Reply:
I assume we need to reload the state and version since it changes frequently across agents.

Reply:
Yes. But with informer, we can sync it on every Add/Modify/Del event, thus we don't need to recalculate it on every request.

Reply:
I think if the load is uniformly distributed among agents, then with >2 agents it might be more effective to use client rather than informer, since specific agent would get changes introduced by all other agents at once, rather than processing several update events?
Also, do you think retry logic would work smoothly with informer, given status update is not using the informer?

Reply:
@annakhm I probably misunderstood the workflow that invokes this function.
I had the feeling we will end up calling this - and retrieving a pool object - every time we need to allocate an IP.
I would think that in this case the frequency would be much higher than CUD operations on the CR.